adapters/southafrica.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * This code is responsible for implementing all methods related to fetching
+ * and returning data for the South African data sources.
+ */
+'use strict';
+
+import { REQUEST_TIMEOUT } from '../lib/constants';
+import { default as baseRequest } from 'request';
+import { default as moment } from 'moment-timezone';
+import _ from 'lodash';
+import { unifyMeasurementUnits, removeUnwantedParameters, unifyParameters } from '../lib/utils';
+const request = baseRequest.defaults({timeout: REQUEST_TIMEOUT});
+
+exports.name = 'southafrica';
+
+/**
+ * Fetches the data for a given source and returns an appropriate object
+ * @param {object} source A valid source object
+ * @param {function} cb A callback of the form cb(err, data)
+ */
+exports.fetchData = function (source, cb) {
+  request(source.url, function (err, res, body) {
+    if (err || res.statusCode !== 200) {
+      return cb({message: 'Failure to load data url.'});
+    }
+
+    // Wrap everything in a try/catch in case something goes wrong
+    try {
+      // Format the data
+      const data = formatData(body);
+      // Make sure the data is valid
+      if (data === undefined) {
+        return cb({message: 'Failure to parse data.'});
+      }
+      cb(null, data);
+    } catch (e) {
+      return cb({message: 'Unknown adapter error.'});
+    }
+  });
+};
+
+/**
+ * Given fetched data, turn it into a format our system can use.
+ * @param {array} results Fetched source data and other metadata
+ * @return {object} Parsed and standarized data our system can use
+ */
+const formatData = function (data) {
+  // Wrap the JSON.parse() in a try/catch in case it fails
+  try {
+    data = JSON.parse(data);
+  } catch (e) {
+    // Return undefined to be caught elsewhere
+    return undefined;
+  }
+  /**
+   * Given a measurement object, convert to system appropriate times.
+   * @param {object} m A source measurement object
+   * @return {object} An object containing both UTC and local times
+   */
+  var parseDate = function (m) {
+    var date = moment.tz(m, 'YYYY-MM-DDHH:mm', 'Africa/Johannesburg');
+    return {utc: date.toDate(), local: date.format()};
+  };
+  var measurements = [];
+  _.forEach(data, function (s) {
+    const base = {
+      location: s.location,
+      city: s.city,
+      coordinates: {
+        latitude: Number(s.latitude),
+        longitude: Number(s.longitude)
+      },
+      attribution: [{name: 'South African Air Quality Information System', url: 'http://saaqis.environment.gov.za'}],
+      averagingPeriod: {unit: 'hours', value: 1}
+    };
+    _.forOwn(s.monitors, function (v, key) {
+      if (v.value !== null) {
+        var m = _.clone(base);
+        m.parameter = v.Pollutantname;
+        m.value = Number(v.value);
+        m.unit = v.unit;
+        m.date = parseDate(v.DateVal);
```

Comment:
Looks like sometimes `dateVal` comes back `null`. For example: 
```
{"IP":null,"display":true,"STA_O3Factor":1.0,"STA_O3BFactor":0.0,"serialCode":304,"name":"Driftsands","location":"Driftsands Nature Reserve, Delft, Cape Town","city":"City of Cape Town","StGroup":null,"StGroupName":"Western Cape ","height":"37","latitude":"-33.991842","longitude":"18.653067","IndexColor":null,"IndexIcon":null,"IndexName":null,"environment":"N/A","GIS_Monitors":null,"GIS_channel":null,"GIS_Units":null,"GIS_LastRecivedValues":null,"timebase":60,"owner":"Western Cape Province","IndexValue":null,"Active":1,"UseMapView":1,"FileName":null,"ImageDescription":null,"ImageStatus":null,"NonContinuous":false,"AQSCODE":null,"STA_ImageData":null,"STA_Notes":"N/A","DateVal":null,"IndexValueDateTime":null,"STA_Target":"Urban","CBO_ItemIndex":2,"TargetId":18,"asix":null,"dateDisplay":"05/04/2021 10:00","monitors":[{"channel":2,"type":46,"value":"12.1","Lastvalue":"12.1","state":1,"name":"NO2","unit":"ppb","stationSerialCode":304,"stationName":"Driftsands","owner":"Western Cape Province","Color":null,"IndexIcon":null,"indexVal":0,"TimeBase":0,"StationTimeBase":60,"DateVal":null,"NumericFormat":5.3,"Pollutantname":"NO2","Description":"","asix":null,"isPolPos":false,"PollutantTimeBase":0,"PollutantID":1,"PctValid":75,"Active":1,"UseMapView":1,"isIndex":true,"indexName":null,"StationTag":"","ShortName":"Driftsands","County":"Western Cape"
```
On the website, it just reports it as `NO_REPORT_DATA` for the whole station so I would filter out any measurements that have null Date values